invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+	}
+
+	toSign := append([]byte(hrp), taggedFieldsBytes...)
+	hash := chainhash.HashB(toSign)
+
+	// We use compact signature format, and also encoded the recovery ID
+	// such that a reader of the invoice can recover our pubkey from the
+	// signature.
+	sign, err := signer.SignCompact(hash)
+	if err != nil {
+		return "", err
+	}
+
+	// From the header byte we can extract the recovery ID, and the last 64
+	// bytes encode the signature.
+	recoveryID := sign[0] - 27 - 4
+	var sigBytes [64]byte
+	copy(sigBytes[:], sign[1:])
+
+	// If the pubkey field was explicitly set, it must be set to the pubkey
+	// used to create the signature.
+	if invoice.PubKey != nil {
+		var signature *btcec.Signature
+		lnwire.DeserializeSigFromWire(&signature, sigBytes)
+		valid := signature.Verify(hash, invoice.PubKey)
+		if !valid {
+			return "", fmt.Errorf("signature does not match " +
+				"provided pubkey")
+		}
+	}
+
+	// Convert the signature to base32 before writing it to the buffer.
+	signBase32, err := bech32.ConvertBits(append(sigBytes[:], recoveryID), 8, 5, true)
+	if err != nil {
+		return "", err
+	}
+	bufferBase32.Write(signBase32)
+
+	// Now we can create the bech32 encoded string from the base32 buffer.
+	b32, err := bech32.Encode(hrp, bufferBase32.Bytes())
+	if err != nil {
+		return "", err
+	}
+
+	return b32, nil
+}
+
+// validateInvoice does a sanity check of the provided Invoice, making sure it
+// has all the necessary fields set for it to be considered valid by BOLT-0011.
+func validateInvoice(invoice *Invoice) error {
+	// The net must be set.
+	if invoice.Net == nil {
+		return fmt.Errorf("net params not set")
+	}
+
+	// The invoice must contain a payment hash.
+	if invoice.PaymentHash == nil {
+		return fmt.Errorf("no payment hash found")
+	}
+
+	// Either Description or DescriptionHash must be set, not both.
+	if invoice.Description != nil && invoice.DescriptionHash != nil {
+		return fmt.Errorf("both description and description hash set")
+	}
+	if invoice.Description == nil && invoice.DescriptionHash == nil {
+		return fmt.Errorf("neither description nor description hash set")
+	}
+
+	// Check that we support the field lengths.
+	if len(invoice.PaymentHash) != 32 {
+		return fmt.Errorf("unsupported payment hash length: %d",
+			len(invoice.PaymentHash))
+	}
+
+	if invoice.DescriptionHash != nil && len(invoice.DescriptionHash) != 32 {
+		return fmt.Errorf("unsupported description hash length: %d",
+			len(invoice.DescriptionHash))
+	}
+
+	if invoice.PubKey != nil &&
+		len(invoice.PubKey.SerializeCompressed()) != 33 {
+		return fmt.Errorf("unsupported pubkey length: %d",
+			len(invoice.PubKey.SerializeCompressed()))
+	}
+
+	return nil
+}
+
+// parseData parses the data part of the invoice. It expects base32 data
+// returned from the bech32.Decode method, except signature.
+func parseData(invoice *Invoice, data []byte, net *chaincfg.Params) error {
+	// It must contain the timestamp, encoded using 35 bits (7 groups).
+	if len(data) < 7 {
```

Comment:
Should be using the `base5timeStampGroup` (or w/e the constant is called :p) constant here, and below in this function.